### ISSUE TITLE
Fine Tune Temperature PID Controller

### ISSUE THREAD

**Issue by kraemr** [qitechgmbh/control#174]

**Comment by kraemr** [comment]
Currently the Temperature PID Controller Overshoots then after some time its hovering around -1 to -3 of difference to the target temperature

**Comment by 1cedsoda** [comment]
[I invented better temperature controller](https://github.com/qitechgmbh/notebooks/blob/main/heating/controller_pdc_gd.ipynb)
It's a semiadaptive **PDC** controller.

# Semiadpative PDC Controller

It uses $K_p$ & $K_d$ normally, but instead of $K_i$ to extinguish the steady state error we use a constant gain $K_c$ which is adapted as the controller runs.

Simulation says it should be stable ($\pm0.5 \degree C$) around 3x faster than a comparable PID controller. It can have near zero oscillation given a low enough $K_P$ and $\eta_c$.

## Signal
The signal of the controller should clamped to $[0, 1]$ which is the power rate for the heating element.

$signal = K_p \cdot error_{t_i} + K_d \cdot (error_{t_i} - error_{t_{i-1}}) + K_c$

## Error Function

$error = T_{Target} - T_{Current}$

## Parameters
- $K_p$ Proportional gain (hardcoded)
- $K_d$ Derivative gain (hardcoded)
- $\eta_c$ Learning rate for gradient descent of $K_c$
- $\tau$ Learning time interval

## Gradient Descent for $K_c$
$K_c$ is the constant signal gain.

The $K_c$ is re-calculated every $\tau$ seconds using gradient descent.

$K_c$ is should be clamped to $[0,1]$

The cost function is a *Mean Signed Square Error* function.
Its applied to the last error in the $\tau$ interval.

$\text{MSSE} = \frac{\sum_{t \in \tau}|error_t| \cdot error_i}{N_{\tau}}$

## Tuning
1. First tune $K_p$ that it overshoots the target temp, and then settles into an equilibrium below the target temperature without oscillation.

2. Then tune $\eta_c$ that after the overshoot there is an even smaller undershoot but then equilibrium is reached (without another overshoot) (this can better be seen by watching $K_c$ value itself)

3. Increase $K_d$ to reduce overshoot and reach the equilibrium faster.

$K_p$ and $\eta_c$ can be tuned live for oscillation reduction,  but everything else needs a cold start.

There will most likely always be overshoot if not operating near the max temperature of the system. Overshoot is naturally bigger for smaller target temperatures.

**Comment by kraemr** [comment]
Not Planned for now
